src/models/StyledComponent.js

```diff
@@ -60,13 +61,53 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
     static componentStyle: Object
     static warnTooManyClasses: Function
 
+    static getDerivedStateFromProps(
+      nextProps: { theme?: Theme, [key: string]: any },
+      prevState
+    ) {
+      // If this is a staticaly-styled component, we don't need to listen to
+      // props changes to update styles
+      if (prevState.isStatic) {
+        return null
+      }
+
+      const theme = determineTheme(
+        nextProps,
+        prevState.theme,
+        prevState.defaultProps
+      )
+      const generatedClassName = prevState.generateClassName(theme, nextProps)
+
+      return { theme, generatedClassName }
+    }
+
     attrs = {}
-    state = {
-      theme: null,
-      generatedClassName: '',
+    state: {
+      theme?: any,
+      defaultProps: any,
+      generateClassName: (theme: any, props: any) => any,
+      isStatic: boolean,
+      generatedClassName: string,
     }
     unsubscribeId: number = -1
 
+    constructor(...args) {
+      super(...args)
+
+      this.state = {
+        ...this.getInitialState(),
+        defaultProps: this.constructor.defaultProps,
+        generateClassName: this.generateClassName.bind(this),
+        isStatic: this.constructor.componentStyle.isStatic,
+      }
+
+      // We need to inject styles in SSR
+      const styleSheet = this.context[CONTEXT_KEY] || StyleSheet.master
+      if (!IS_BROWSER || styleSheet.forceServer) {
+        this.injectStyles()
```

Comment:
This has the same problem as gDSFP since the constructor also need to be pure.

So with things like suspense on the server, this will get called multiple times.

Reply:
How can we make sure to play nicely with side effects right now before the RFC goes in? I realise our logic here is quite specific.

The StyleSheet can handle multiple injection (caching prevents it on multiple levels to happen anyway)

Also, what do you geese think of abstracting this as a simple reducer-like fn so that we can reuse it more easily without a component (might be worth replacing ComponentStyle in the process)

Reply:
Also, AFAIK even if the constructor is called, the component might not mount.